mac: Allow configuring usage of FullMemoryWE (fixes #70)

david-sawatzke opened the pull request:
On ecp5 `FullMemoryWE` leads to an increase of DP16KD block mem, while
it works better on Intel/Altera devices according to
6c3af746e28f55089c9f79966205499af394ae34.

Simple solution: Make it configurable

EDIT: Original issue leading to `FullMemoryWE` being used is #63

david-sawatzke committed:
mac: Allow configuring usage of FullMemoryWE (fixes #70)

On ecp5 `FullMemoryWE` leads to an increase of DP16KD block mem, while
it works better on Intel/Altera devices according to
6c3af746e28f55089c9f79966205499af394ae34.

Simple solution: Make it configurable

Disasm commented:
Maybe this wrapper can be enabled based on the platform used. The wrapper is not documented so it's not clear what it does, it will be difficult for others to decide whether they should turn this setting on or not.

rowanG077 commented:
Please if you add this config options also add it to `gen.py` could be done by platform as well.

david-sawatzke commented:
> Maybe this wrapper can be enabled based on the platform used

Is there any easy way to query this?

Since this also an issue [here](https://github.com/enjoy-digital/litex/blob/79ac09316ae00714f8d53d91bb7ab3419f2f74b2/litex/soc/integration/soc.py#L1243) maybe the automatic selection should be done in litex or at least some infra? Maybe a new `PartialWE` wrapper that selects either nothing or FullMemWE depending on the Platform (or maybe even depending on the toolchain)?

david-sawatzke committed:
liteeth/core: Allow configuration of full_mem_we parameter

Disasm commented:
@david-sawatzke hmm, probably there is no simple way to query the platform here. Anyway, your idea with a platform-dependent wrapper looks really good and it should be easy to implement.

david-sawatzke commented:
@rowanG077 I've added a commit to make this configurable in `gen.py`, but it's untested. Seems to be obviously correct, but could you please check?

david-sawatzke commented:
> should be easy to implement.

No idea how to even approach that. For now I'm happy with leaving a magic/more magic knob. Maybe a comment that makes it clear that this depends on the platform and toolchain with obvious choices for yosys or Intel/Altera?

Disasm commented:
LiteX uses "special overrides" to implement platform-specific things DDR and tristates: https://github.com/enjoy-digital/litex/blob/master/litex/build/lattice/common.py#L441-L450 Hopefully this wrapper can be implemented in the same way.
But I agree, probably this switch is good enough solution for now.

david-sawatzke commented:
@disasm I've added a nice comment, which is probably the most documentation FullMemoryWE has ever seen and adjusted the parameter name & implementation to work more like the similar parameter in [`add_sdram`](https://github.com/enjoy-digital/litex/blob/79ac09316ae00714f8d53d91bb7ab3419f2f74b2/litex/soc/integration/soc.py#L1383) in litex.

enjoy-digital commented:
Thanks @david-sawatzke for identifying the issue on ECP5 and providing a solution! As you discussed all together the current infrastructure does not currently allow simple specialization of memories. That's one of the many points I'd like to improve while reworking a bit the infrastructure, so for now your solution is good enough and we'll improve it in the future to avoid having to set this manually.
